text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,414 @@
+
+#### Functions/methods
+
+Whenever a function/method is present, we will generate a link to it. Example:
+
+```rust
+let x: Fn() = some_fn; // We will link to `some_fn`.
+some_fn(); // We will link to `some_fn`.
+ty.a().b(); // We will link to `a` and to `b`.
+```
+
+#### Macros
+
+Macro calls will get a link:
+
+```rust
+some_macro!(); // Will link to `some_macro!`.
+```
+
+#### Constants/statics
+
+Both these kinds of item will get a link too:
+
+```rust
+static FOO: u32 = 0;
+const BAR: u32 = 0;
+
+let x = FOO; // `FOO` gets a link.
+let y = BAR; // `BAR` gets a link.
+```
+
+#### Modules
+
+Inline modules will never generate a "jump to definition" link, only a "jump to documentation" one. However, other modules will behave as follows: if they are private and the `--document-private-items` is not in use, they will generate a "jump to definition" link. Otherwise they will generate a "jump to documentation" link.
+
+#### Paths
+
+Any part of a path which isn't a keyword will get a link. So for example:
+
+```rust
+use crate::foo::bar{self, Type};
+```
+
+In this one, `foo`, `bar` and `Type` will get a link.
+
+```rust
+let x = X::Y::Z;
+```
+
+In this one, `X`, `Y` and `Z` will get a link.
+
+## Cross-crate links
+
+Some issues that can appear when generate cross-crate links is to link to non-existent locations. For example, if a version your crate depends on when published is yanked, your links (not just source code links but also links in documentation) will target non-existent pages and will result in 404 or "file not found" if local.
+
+A discussion about this suggested to instead add a new argument in the URL (`?source=true`) which would automatically click on the source link of the item on its documentation page. The problem is that private elements (unless the documentation is generated with `--document-private-items`) don't have a documentation page, which reduces the possibilities of this solution quite a lot. This fix would also only worked on web platform implementing it as it requires a server to pick a compatible version (in the semver meaning) for this crate. Another problem is: what to do if there is no compatible version (in the semver meaning still) available? For example, if there was only a `0.15` version published, `0.14.x` and `0.16.y` are not compatible to it. Then it would very likely just don't redirect and leave the user on the 404 page.
+
+The idea here would be to follow the current URL strategy in use for cross-crate items: generate a URL relative to the root path. If `--extern-html-root-url` isn't used, then the root-path is the one provided to rustdoc where to build documentation.
+
+This is somewhat the same answer for `cargo --no-deps`: unless rustdoc actually checks if the dependencies' locations are empty or not, there is no way for rustdoc to know if they're present. As such, it should simply assume that they are present and generate links for them as it does for the foreign items in the documentation pages.
```

Comment:
If I understand the whole *cross-crate links* section correctly, you seem to be saying that it is not possible at all for rustdoc to detect if a to-be-generated link would point to a nonexistent location.

However, from my limited experience I've gathered from writing several <kbd>A-cross-crate-reexport</kbd> PRs, I've learned that rustdoc *does* know if a cross-crate link does not resolve. Maybe that's not always the case but at least it is for code using [`html::format::href()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/html/format/fn.href.html).

So from what I know, rustdoc should be able to “gray out” / not generate (most? all?) links that would be dead. Specifically, the following paragraph should be updated to mention this:

> As such, it should simply assume that they are present and generate links for them as it does for the foreign items in the documentation pages.

Am I missing something major?

Reply:
I tried to mean not the cross-crate item itself but the actual HTML file destination. As I mentioned, you can run `cargo --no-deps` and in this case, only your crate will be documented and not its dependencies. As such, any link to any dependency will result in a 404. Does it clear the confusion? How could I rewrite it to make it less confusing?

Reply:
I tried to improve the wording and explanations of this section.